velox/functions/lib/string/StringImpl.h

```diff
@@ -653,4 +653,84 @@ FOLLY_ALWAYS_INLINE void pad(
       padPrefixByteLength);
 }
 
+FOLLY_ALWAYS_INLINE static void
+initCapAscii(char* output, const char* input, size_t length) {
+  bool newWord = true;
+  for (size_t i = 0; i < length; ++i) {
+    if (std::isspace(input[i])) {
+      output[i] = input[i];
+      newWord = true;
+    } else {
+      output[i] = newWord ? std::toupper(input[i]) : std::tolower(input[i]);
+      newWord = false;
+    }
+  }
+}
+
+FOLLY_ALWAYS_INLINE size_t initCapUnicode(
+    char* output,
+    size_t outputLength,
+    const char* input,
+    size_t inputLength) {
+  auto inputIdx = 0;
+  auto outputIdx = 0;
+  bool newWord = true;
+
+  while (inputIdx < inputLength) {
+    int size;
+    utf8proc_int32_t nextCodePoint =
+        utf8proc_codepoint(&input[inputIdx], input + inputLength, size);
+    if (UNLIKELY(nextCodePoint == -1)) {
+      // Invalid input string, copy the remaining of the input string as is to
+      // the output.
+      std::memcpy(&output[outputIdx], &input[inputIdx], inputLength - inputIdx);
+      outputIdx += inputLength - inputIdx;
+      return outputIdx;
+    }
+
+    inputIdx += size;
+    utf8proc_int32_t modifiedCodePoint = newWord
+        ? utf8proc_toupper(nextCodePoint)
+        : utf8proc_tolower(nextCodePoint);
+    newWord = utf8proc_category(nextCodePoint) ==
+        UTF8PROC_CATEGORY_ZS; // Check if it's a space character.
+
+    VELOX_USER_CHECK_LT(
+        outputIdx + utf8proc_codepoint_length(modifiedCodePoint),
+	outputLength,
+	"Access out of bound, index {}",
+	outputIdx);
+
+    auto newSize = utf8proc_encode_char(
+        modifiedCodePoint,
+        reinterpret_cast<unsigned char*>(&output[outputIdx]));
+    outputIdx += newSize;
+  }
+  return outputIdx;
+}
+
+template <bool ascii, typename TOutString, typename TInString>
+FOLLY_ALWAYS_INLINE bool initCap(TOutString& output, const TInString& input) {
+  if constexpr (ascii) {
+    // If the input string is ASCII, resize the output string to the same size as the input string.
+    output.resize(input.size());
+    initCapAscii(output.data(), input.data(), input.size());
+  } else {
+    // If the input string is Unicode, resize the output string to 4 times the size of the input string.
+    // This is done to accommodate the worst-case scenario where each Unicode character takes up 4 bytes.
+    output.resize(input.size() * 4);
+    auto size = initCapUnicode(
+        output.data(), output.size(), input.data(), input.size());
+    output.resize(size);
+  }
+  return true;
+}
+
+// Inplace ascii initCap.
+template <typename T>
+FOLLY_ALWAYS_INLINE bool initCapAsciiInPlace(T& str) {
+  initCapAscii(str.data(), str.data(), str.size());
```

Comment:
Coding style says "do not abbreviate". Let's come up with a more descriptive name to use in Velox implementation than 'initcap'.

Reply:
Is initialCapitalization better? No one in the SQL / spark world seems to call it that but happy to change it you want?

Reply:
I didn't know that 'initcap' is some sort of a known "word". I see now that many databases implement 'initcap' function: Oracle, Teradata, Postgres, Redshift. We can keep this as is.

I'm seeing that some of the document this function nicely. We can use these as an inspiration for our docs.

Redshift: https://docs.aws.amazon.com/redshift/latest/dg/r_INITCAP.html

```
Capitalizes the first letter of each word in a specified string. INITCAP supports UTF-8 multibyte characters, up to a maximum of four bytes per character.

The INITCAP function makes the first letter of each word in a string uppercase, and any subsequent letters are made (or left) lowercase. Therefore, it is important to understand which characters (other than space characters) function as word separators. A word separator character is any non-alphanumeric character, including punctuation marks, symbols, and control characters. All of the following characters are word separators:


! " # $ % & ' ( ) * + , - . / : ; < = > ? @ [ \ ] ^ _ ` { | } ~ 

Tabs, newline characters, form feeds, line feeds, and carriage returns are also word separators.
```

Informix: https://www.ibm.com/docs/it/informix-servers/12.10?topic=smf-initcap-function

```
Use the INITCAP function to replace the first letter of every word in a character string with an uppercase letter. The INITCAP function assumes a new word whenever the function encounters a letter that is preceded by any character other than a letter.
```